MP/Add business table, model, and associations

MPucheta opened the pull request.

MPucheta committed:
Add business table, model, and assocciations

abelosorio commented on src/migrations/20201019191554-add-business.js:
```diff
@@ -0,0 +1,35 @@
+        primaryKey: true,
+        defaultValue: Sequelize.UUIDV4
+      },
+      fantasyName: {
```
Unique?

abelosorio commented on src/migrations/20201019191554-add-business.js:
```diff
@@ -0,0 +1,35 @@
+        type: Sequelize.STRING,
+        allowNull: false
+      },
+      legalInformationId: {
```
The FK should be in the LegalInformation table ("LegalInformation belongs to Business", "Business has 1 LegalInformation").

abelosorio commented on src/migrations/20201019192018-add-businessPhysicalAddress.js:
```diff
@@ -0,0 +1,38 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.createTable('businessPhysicalAddress', {
```
Can the same address belong to multiple businesses? This relationship table should exist only for n:m relationships.

abelosorio commented on src/models/business.js:
```diff
@@ -0,0 +1,25 @@
+  })
+
+  Business.associate = (models) => {
+    Business.belongsTo(models.legalInformation)
```
Business.hasOne(models.legalInformation).

abelosorio commented on src/models/business.js:
```diff
@@ -0,0 +1,25 @@
+  Business.associate = (models) => {
+    Business.belongsTo(models.legalInformation)
+    Business.hasOne(models.user)
+    Business.belongsToMany(models.physicalAddress, {
```
Maybe there is a requirement I don't know about, but this doesn't make much sense to me. I'd have written this as "Business.hasMany(models.physicalAddress)".

abelosorio commented on src/models/legalInformation.js:
```diff
@@ -22,16 +22,13 @@ export default (sequelize, DataTypes) => {
   })
 
+  LegalInformation.associate = (models) => {
+    LegalInformation.hasOne(models.business)
```
LegalInformation.belongsTo(model.business)
LegalInformation.hasOne(models.physicalAddress)

abelosorio commented on src/migrations/20201019203657-add-businessId-to-user.js:
```diff
@@ -0,0 +1,16 @@
+module.exports = {
+  up: (queryInterface, Sequelize) => {
+    return queryInterface.addColumn('user', 'businessId', {
```
This requires you to create the business **first**, then create a user associating it to the business you just created. I don't think that's a business logic you want to have in the backend.

nahudalla replied:
Remove this migration, you already created an association table.

MPucheta committed:
Reimplement business relationships and add constraints.

This includes replacing the address N:M relationship to be 1:1, replacing the legalInformation to be N:0, and adding de userBusiness table.

The constraint is a fantasyName check, not allowing duplicates

abelosorio commented on src/migrations/20201019191554-add-business.js:
```diff
@@ -1,35 +1,54 @@
+      }, { transaction })
+
+      await queryInterface.addIndex('business', [queryInterface.sequelize.fn('lower', queryInterface.sequelize.col('fantasyName'))], {
+        name: 'business_fantasyName_unique_case_insensitive',
```
_idx

ref: https://gist.github.com/popravich/d6816ef1653329fb1745

nahudalla commented on src/migrations/20201019191554-add-business.js:
```diff
@@ -0,0 +1,54 @@
+  },
+
+  down: (queryInterface, Sequelize) => {
+    return queryInterface.sequelize.transaction(async (transaction) => {
+      await queryInterface.dropTable('business', { transaction })
+      await queryInterface.removeIndex('business', 'business_fantasyName_unique_case_insensitive', { transaction })
+    })
```
Did this work when you tested it? It think you need remove the index first and then the table :thinking: If it worked, then the index is probably removed automatically.

abelosorio replied:
Maybe "removeIndex" does a "DROP IF EXISTS"

nahudalla commented on src/models/business.js:
```diff
@@ -0,0 +1,25 @@
+
+  Business.associate = (models) => {
+    Business.belongsTo(models.legalInformation)
+    Business.belongsTo(models.physicalAddress, { as: 'realAddress' })
```
I think you need to specify the field to use for the association. I think changing the name of the association to `realAddress` will make Sequelize to expect a `realAddressId` to be present in the `Business` model instead of the default `physicalAddressId`.

MPucheta committed:
Add missing foreign key in association and rename business index to have suffix _idx
